packages/pwa-kit-runtime/src/utils/logger.js

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+const LOG_LEVELS = ['debug', 'info', 'warn', 'error']
+const DEFAULT_LOG_LEVEL = 'info'
+const DEFAULT_LOG_FORMAT = 'JSON'
+
+const isServerSide = typeof window === 'undefined'
+
+const LOG_LEVEL = isServerSide
+    ? process.env.PWAKIT_LOG_LEVEL || DEFAULT_LOG_LEVEL
+    : DEFAULT_LOG_LEVEL
+
+const LOG_FORMAT = isServerSide
+    ? process.env.PWAKIT_LOG_FORMAT || DEFAULT_LOG_FORMAT
+    : DEFAULT_LOG_FORMAT
+
+/**
+ * The PWAKITLogger provides structured logging with different log levels.
+ *
+ */
+export class PWAKITLogger {
+    /**
+     * Creates an instance of PWAKITLogger.
+     * @param {string} logLevel - The log level to set for the logger.
+     * @param {string} [format=DEFAULT_LOG_FORMAT] - The format in which to print log messages. Can be 'JSON' or 'TEXT'.
+     * @param {string} packageName - The name of the package where the logger is used.
+     */
+    constructor(logLevel = LOG_LEVEL, format = DEFAULT_LOG_FORMAT, packageName = '') {
+        this.logLevel = LOG_LEVELS.includes(logLevel) ? logLevel : DEFAULT_LOG_LEVEL
+        this.format = format.toUpperCase()
+        this.packageName = packageName
+    }
+
+    /**
+     * Determines if the message should be logged based on the PWAKIT_LOG_LEVEL value.
+     * @param {string} level - The log level of the message.
+     * @returns {boolean} - Returns true if the message should be logged, otherwise false.
+     */
+    shouldLog(level) {
+        return LOG_LEVELS.indexOf(level) >= LOG_LEVELS.indexOf(this.logLevel)
+    }
+
+    /**
+     * Formats the log message based on the provided context and format.
+     *
+     * @param {string} message - The log message.
+     * @param {Object} details - The context for the log message.
+     * @param {string} details.namespace - The namespace 'packageName.component.action' of the log message.
+     * @param {Object} details.additionalProperties - Additional properties to include in the log message.
+     * @returns {Object|string} - The formatted log message, either as a JSON object or a string.
+     */
+    formatLogMessage(message, {namespace, additionalProperties}) {
+        const logMessage = {
+            namespace: `${this.packageName}.${namespace}`,
+            message: message.trim(),
+            ...(additionalProperties && {additionalProperties})
+        }
+
+        if (this.format === 'TEXT') {
+            return JSON.stringify(logMessage)
+        }
```

Comment:
Is this common that the "text" version of a message is the literal stringified value of the data?

Reply:
I've reformatted the TEXT version of the message to make it easier to read.